pkg/imager/cache/cache.go

```diff
@@ -0,0 +1,183 @@
+func Generate(images []string, platform string, insecure bool, dest string) error {
+	v1Platform, err := v1.ParsePlatform(platform)
+	if err != nil {
+		return fmt.Errorf("parsing platform: %w", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "talos-image-cache-gen")
+	if err != nil {
+		return fmt.Errorf("creating temporary directory: %w", err)
+	}
+
+	defer os.RemoveAll(tmpDir) //nolint:errcheck
+
+	if err := os.MkdirAll(filepath.Join(tmpDir, blobsDir), 0o755); err != nil {
+		return err
+	}
+
+	nameOptions := []name.Option{
+		name.StrictValidation,
+	}
+
+	craneOpts := []crane.Option{
+		crane.WithAuthFromKeychain(
+			authn.NewMultiKeychain(
+				authn.DefaultKeychain,
+				github.Keychain,
+			),
+		),
+	}
+
+	remoteOpts := []remote.Option{
+		remote.WithAuthFromKeychain(authn.NewMultiKeychain(
+			authn.DefaultKeychain,
+			github.Keychain,
+		)),
+		remote.WithPlatform(*v1Platform),
+	}
+
+	if insecure {
+		craneOpts = append(craneOpts, crane.Insecure)
+		nameOptions = append(nameOptions, name.Insecure)
+	}
+
+	for _, src := range images {
+		ref, err := name.ParseReference(src, nameOptions...)
+		if err != nil {
+			return fmt.Errorf("parsing reference %q: %w", src, err)
+		}
+
+		if err := os.MkdirAll(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "reference"), 0o755); err != nil {
+			return err
+		}
+
+		if err := os.MkdirAll(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "digest"), 0o755); err != nil {
+			return err
+		}
+
+		manifest, err := crane.Manifest(
+			src,
+			craneOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("fetching manifest %q: %w", src, err)
+		}
+
+		rmt, err := remote.Get(
+			ref,
+			remoteOpts...,
+		)
+		if err != nil {
+			return fmt.Errorf("fetching image %q: %w", src, err)
+		}
+
+		if !strings.HasPrefix(ref.Identifier(), "sha256:") {
+			if err := os.WriteFile(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "reference", ref.Identifier()), manifest, 0o644); err != nil {
+				return err
+			}
+		}
+
+		if err := os.WriteFile(filepath.Join(tmpDir, manifestsDir, ref.Context().RegistryStr(), ref.Context().RepositoryStr(), "digest", rmt.Digest.String()), manifest, 0o644); err != nil {
+			return err
+		}
+
+		img, err := rmt.Image()
+		if err != nil {
+			return fmt.Errorf("converting image to index: %w", err)
+		}
+
+		layers, err := img.Layers()
+		if err != nil {
+			return fmt.Errorf("getting image layers: %w", err)
+		}
+
+		for _, layer := range layers {
+			digest, err := layer.Digest()
+			if err != nil {
+				return fmt.Errorf("getting layer digest: %w", err)
+			}
+
+			reader, err := layer.Compressed()
+			if err != nil {
+				return fmt.Errorf("getting layer reader: %w", err)
+			}
+
+			file, err := os.Create(filepath.Join(tmpDir, blobsDir, digest.String()))
+			if err != nil {
+				return err
+			}
+
+			defer file.Close() //nolint:errcheck
```

Comment:
`defer`s are function scoped, not lexical, so they'll get executed after all loops. I don't think you wont this here.

Reply:
Won't this be like an automatic closure here? Because file variable is also scoped in this iteration of the loop, it would probably be replicated for each iteration? Or not, and it produces a single deferred call after the last iteration with the last value of the variable?

Reply:
It will produce defers for each file variable individually, but it will not execute them until function finishes.

Reply:
Yes, this is what I also supposed would happen. Thanks for clarification